### DISCUSSION TITLE
Making shapely.wkt.loads() able to load TIN and POLYHEDRALSURFACE and write them to their corresponding data types in PostGIS

### DISCUSSION THREAD

**Discussion by swiss-knight** [shapely/shapely#1369] [Ideas]
Hi,

I was naively giving Shapely a try for loading `TIN` and `POLYHEDRALSURFACE`, two data types that are available to PostGIS by the mean of [SFCGAL](https://oslandia.gitlab.io/SFCGAL/): https://postgis.net/workshops/postgis-intro/3d.html 
My dream is to be able to load them in order to write them programmatically to a PostGIS database where some adequate fields are waiting for them. Of course I can still manage to write the raw SQL query using `psycopg2` for example, but it could be nice if Shapely could handle those data types.

Here's my attempt with a `POLYHEDRALSURFACE`:
```
In [1]: from shapely import wkt
In [2]: wkt.loads('POLYHEDRALSURFACE Z (((0 0 0,1 1 1,1 2 1,0 0 0)),((0 0 0,1 2 1,3 2 4,0 0 0)),((0 0 0,3 2 4,2 2 1,0 0 0)),((0 0 0,2 2 1,1 1 1,0 0 0)))')
ParseException: Unknown type: 'POLYHEDRALSURFACE'
Traceback (most recent call last):

  File "/tmp/ipykernel_1135764/1540595037.py", line 1, in <cell line: 1>
    wkt.loads('POLYHEDRALSURFACE Z (((0 0 0,1 1 1,1 2 1,0 0 0)),((0 0 0,1 2 1,3 2 4,0 0 0)),((0 0 0,3 2 4,2 2 1,0 0 0)),((0 0 0,2 2 1,1 1 1,0 0 0)))')

  File "/usr/local/lib/python3.10/dist-packages/shapely/wkt.py", line 22, in loads
    return geos.WKTReader(geos.lgeos).read(data)

  File "/usr/local/lib/python3.10/dist-packages/shapely/geos.py", line 325, in read
    raise WKTReadingError(

WKTReadingError: Could not create geometry because of errors while reading input.
```

and with the corresponding `TIN`:
```
In [1]: from shapely import wkt
In [2]: wkt.loads('TIN Z (((0 0 0,1 1 1,1 2 1,0 0 0)),((0 0 0,1 2 1,3 2 4,0 0 0)),((0 0 0,3 2 4,2 2 1,0 0 0)),((0 0 0,2 2 1,1 1 1,0 0 0)))')
ParseException: Unknown type: 'TIN'
Traceback (most recent call last):

  File "/tmp/ipykernel_1135764/3725657591.py", line 1, in <cell line: 1>
    wkt.loads('TIN Z (((0 0 0,1 1 1,1 2 1,0 0 0)),((0 0 0,1 2 1,3 2 4,0 0 0)),((0 0 0,3 2 4,2 2 1,0 0 0)),((0 0 0,2 2 1,1 1 1,0 0 0)))')

  File "/usr/local/lib/python3.10/dist-packages/shapely/wkt.py", line 22, in loads
    return geos.WKTReader(geos.lgeos).read(data)

  File "/usr/local/lib/python3.10/dist-packages/shapely/geos.py", line 325, in read
    raise WKTReadingError(

WKTReadingError: Could not create geometry because of errors while reading input.
```

But with a `MULTIPOLYGON` (the coordinates are always the same) it is obviously OK (even so the resulting multipolygon is not valid, hence the reddish color in my Python IDE, because, well, it's a true 3D object and there are overlaps when seen as a multipolygon...):
```
In [1]: from shapely import wkt
In [2]: wkt.loads('MULTIPOLYGON Z (((0 0 0,1 1 1,1 2 1,0 0 0)),((0 0 0,1 2 1,3 2 4,0 0 0)),((0 0 0,3 2 4,2 2 1,0 0 0)),((0 0 0,2 2 1,1 1 1,0 0 0)))')
```
![image](https://user-images.githubusercontent.com/19967599/164968828-e9fc22ab-6b77-414b-8a9b-d7a801d60eba.png)

[GDAL](https://github.com/OSGeo/gdal) people seems to have addressed this some time ago: https://gdal.org/development/rfc/rfc64_triangle_polyhedralsurface_tin.html
```
IN [1]: from osgeo import ogr

IN [2]: geom = ogr.CreateGeometryFromWkt('POLYHEDRALSURFACE Z (((0 0 0,1 1 1,1 2 1,0 0 0)),((0 0 0,1 2 1,3 2 4,0 0 0)),((0 0 0,3 2 4,2 2 1,0 0 0)),((0 0 0,2 2 1,1 1 1,0 0 0)))')

IN [3]: geom.GetGeometryName()
Out[3]: 'POLYHEDRALSURFACE'

IN [4]: geom.GetGeometryType()
Out[4]: 1015

IN [5]: geom = ogr.CreateGeometryFromWkt('TIN Z (((0 0 0,1 1 1,1 2 1,0 0 0)),((0 0 0,1 2 1,3 2 4,0 0 0)),((0 0 0,3 2 4,2 2 1,0 0 0)),((0 0 0,2 2 1,1 1 1,0 0 0)))')

IN [6]: geom.GetGeometryName()
Out[6]: 'TIN'

IN [7]: geom.GetGeometryType()
Out[7]: 1016
```

Geometry Type numbers are given by this table:
![image](https://user-images.githubusercontent.com/19967599/164970258-a6b0d996-bdbc-4943-beae-445cd313312b.png)
Source: https://gdal.org/development/rfc/rfc64_triangle_polyhedralsurface_tin.html#geometry-types

That would be great if Shapely can load them, and also write them to PostGIS / dump them to text files.    
What do you think?

**Comment by tomplex** [comment]
I'm curious why you want to load these geometries in shapely as a prerequisite to loading them to postgres. shapely does not have any special connector with postgres; even if shapely *could* handle them, you'd just have to dump them right back out to one of WKT, WKB, or GeoJSON and construct a query to load them anyway. 

If what you're after is simply format conversion (e.g. You have WKT and want GeoJSON) then I'd recommend the library geomet, which is focused on that sort of thing. They don't support those types now, but it shouldn't be hard to add (I am a contributor there and could probably knock something out for you if you wanted).

P.S. I'd personally recommend that you create a csv file with your geometries in it at WKT and use `psycopg2.cursor.copy_expert` to copy them to postgres quickly and skip shapely altogether.

**Reply by swiss-knight** [reply]
For the moment I'm doing it in a dirty way:
```
import psycopg2
from shapely import wkt
from shapely.geometry import MultiPolygon

Facets = []
for triangles in verticesArray: # a numpy array of triangles' vertices
    Facets.append(Polygon(triangles))

# the following can be done because the structure of the vertices is exactly the same than for the MultiPolygon
# but the usage of the .replace() method is ugly to me...
surface = wkt.dumps(MultiPolygon(Facets)).replace('MULTIPOLYGON', 'POLYHEDRALSURFACE')

sql = "INSERT INTO mytable (mypolyhedralsurfacefield) VALUES ( ST_GeomFromText(%(surface)s, 4326) ) WHERE id = 1;"
conn = psycopg2.connect(**db_params)
curs = conn.cursor()
curs.execute(sql, vars = {"surface": surface})
```

Something like that... to get the idea.

I don't really know how the best I can upload a TIN or a POLYHEDRALSURFACE into PostGIS from Python (those two special geometry columns already exists in a table).

At the really end I wish I could do a `ST_3Dintersection()` between a polyhedralsurface and a TIN terrain, but I'm just noticing it takes ages (...) for large TINs (>=50'000 triangles)! :'(

**Reply by tomplex** [reply]
given your example, I would recommend you submit a feature request to the [geomet](https://github.com/geomet/geomet) library to support the polyhedralsurface type in their dumps/loads functions. [Here's an example workflow](https://gist.github.com/tomplex/860380e4ef885cff5579cf6c16634bdc) that you could use to create and load many surface geometries quickly.

**Reply by swiss-knight** [reply]
Thanks a lot for your example, which is really interesting; I didn't know about that `psycopg2` [`cursor.copy_expert()`](https://www.psycopg.org/docs/cursor.html#cursor.copy_expert) method. It's much quicker, indeed: I divided the execution time by roughly 55 (!) on a TIN containing ~1e7 triangles.

--> https://github.com/geomet/geomet/issues/80

**Comment by jorisvandenbossche** [comment]
I think @tomplex is giving good advice, but just to give the background on why Shapely itself does not support this: the WKT reader/writer in Shapely is a binding to the underlying GEOS library (which powers all functionality of Shapely). At the moment, GEOS does not support non-standard geometry types such as TIN or POLYHEDRALSURFACE (that might change in the future, but until then, Shapely will also quite likely not support this)

**Reply by swiss-knight** [reply]
Thank you for the clear explanations. Hopefully things will change in GEOS to integrate some other 3D data in the future ;)
